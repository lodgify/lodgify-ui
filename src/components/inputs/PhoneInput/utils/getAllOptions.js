import React, { Fragment } from 'react';

import { getIconOrFlag } from './getIconOrFlag';
import { getDiallingCode } from './getDiallingCode';

const optionCache = {};

/**
 * @param  {Object[]} options
 * @return {Object[]}
 */
export const getAllOptions = options =>
  options.map(({ value, label }) => {
    if (!optionCache[label]) {
      const iconOrFlag = getIconOrFlag(value);

      optionCache[label] = {
        name: label,
        text: iconOrFlag,
        value,
        content: (
          <Fragment>
            {iconOrFlag}
            {`${label} ${getDiallingCode(value)}`}
          </Fragment>
        ),
      };
    }

    return optionCache[label];
  });
