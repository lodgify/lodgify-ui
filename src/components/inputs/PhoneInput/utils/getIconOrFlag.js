import React from 'react';
import { Flag } from 'semantic-ui-react';

import { lowerCase } from 'utils/lower-case';
import { Icon, ICON_NAMES } from 'elements/Icon';

import { VALID_FLAG_NAMES } from './validFlagNames';

/**
 * If country code known, return an appropriate flag component.
 * Default to returning a phone icon.
 * @param  {string} country
 * @return {Object}
 */
export const getIconOrFlag = country => {
  const flagName = lowerCase(country);

  return flagName && VALID_FLAG_NAMES.includes(flagName) ? (
    <Flag name={flagName} />
  ) : (
    <Icon name={ICON_NAMES.PHONE} size="small" />
  );
};
