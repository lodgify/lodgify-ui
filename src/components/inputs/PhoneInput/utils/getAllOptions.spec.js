jest.mock('./getIconOrFlag');

import { getAllOptions } from './getAllOptions';
import { getIconOrFlag } from './getIconOrFlag';

getIconOrFlag.mockImplementation(value => value);

describe('getAllOptions', () => {
  it('should return the right shape', () => {
    const options = [
      { value: 'JP', label: 'jappppan' },
      { value: 'DE', label: 'gerrrmany' },
      { value: 'GB', label: 'uunnniited' },
    ];

    const actual = getAllOptions(options);

    expect(actual).toMatchSnapshot();
  });

  it('should use the `optionCache`', () => {
    const options = [
      { value: 'JP', label: 'jappppan' },
      { value: 'DE', label: 'gerrrmany' },
      { value: 'GB', label: 'uunnniited' },
    ];

    getIconOrFlag.mockClear();
    getAllOptions(options);

    expect(getIconOrFlag).not.toHaveBeenCalled();
  });
});
