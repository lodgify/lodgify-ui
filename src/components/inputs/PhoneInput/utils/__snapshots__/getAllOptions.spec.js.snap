// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAllOptions should return the right shape 1`] = `
Array [
  Object {
    "content": <React.Fragment>
      JP
      jappppan +81
    </React.Fragment>,
    "name": "jappppan",
    "text": "JP",
    "value": "JP",
  },
  Object {
    "content": <React.Fragment>
      DE
      gerrrmany +49
    </React.Fragment>,
    "name": "gerrrmany",
    "text": "DE",
    "value": "DE",
  },
  Object {
    "content": <React.Fragment>
      GB
      uunnniited +44
    </React.Fragment>,
    "name": "uunnniited",
    "text": "GB",
    "value": "GB",
  },
]
`;
